### ISSUE TITLE
np.take_along_axis gives wrong gradient

### ISSUE THREAD

**Issue by fehiepsi** [google/jax#1521]
The following repro script shows the issue:
```
import jax.numpy as np
from jax import value_and_grad

idx = np.repeat(np.arange(3), 10).reshape((30, 1))

def f(x):
    y = x * np.arange(3.).reshape((1, 3))
    return np.take_along_axis(y, idx, -1).sum()

def g(x):
    y = x * np.arange(3.).reshape((1, 3))
    y = np.broadcast_to(y, (30, 3))
    return np.take_along_axis(y, idx, -1).sum()

print(value_and_grad(f)(1.))  # get 30, 0
print(value_and_grad(g)(1.))  # get 30, 30
```

I think `(30, 30)` is the correct answer.

**Comment by mattjj** [comment]
Thanks for catching this.

By the way, I recommend using `check_grad` from `jax.test_util` for checking derivatives (because you don't have to worry about scalar output, and you can check fwd/rev and arbitrary order all at once):

```python
import jax.numpy as np
from jax import value_and_grad
from jax.test_util import check_grads

idx = np.arange(3).reshape(3, 1)

def h(x):
  y = x * np.arange(3.).reshape((1, 3))
  return np.take_along_axis(y, idx, -1)

check_grads(h, (1.,), 1)
```

```
AssertionError:
0.0 !=
9.63398933411
```

**Comment by mattjj** [comment]
It seems `check_grads(h, (1.,), order=1, modes=["fwd"])` succeeds but `check_grads(h, (1.,), order=1, modes=["rev"])` fails. The issue must be with the gather transpose rule (?).

**Comment by mattjj** [comment]
Yet another reproducer, which can give us a bit more insight into what's going wrong:

```python
from functools import partial
import numpy as onp
from jax import vjp
from jax import lax

start_indices = onp.array([[[0, 0]],
                            [[1, 1]]], onp.int32)
dnums = lax.GatherDimensionNumbers(offset_dims=(),
                                    collapsed_slice_dims=(0, 1),
                                    start_index_map=(0, 1))
slice_sizes = (1, 1)

f = partial(lax.gather, start_indices=start_indices, dimension_numbers=dnums,
            slice_sizes=slice_sizes)

_, f_vjp = vjp(f, onp.array([[1., 1.]]))
g = f_vjp(onp.array([[1.], [1.]]))
print(g)
```

The [`lax.gather` transpose rule](https://github.com/google/jax/blob/eae59d0b2c9cc13e0f2b835c4daa6c7be23ba3b2/jax/lax/lax.py#L2894-L2904) is trivial, so it's hard to imagine an error being there.

@hawkinsp had the brilliant suggestion to see if this repros on all backends (since maybe this could have been an XLA:CPU bug). Unfortunately, it does: it fails on GPU and TPU as well.

**Comment by mattjj** [comment]
Here's the VJP jaxpr (slightly simplified by commenting out a `tie_in`):

```
{ lambda b c ;  ; a.
  let d = scatter-add[ updates_shape=(2, 1)
                       update_jaxpr={ lambda  ;  ; a b.
                                      let c = add a b
                                      in [c] }
                       dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0, 1), scatter_dims_to_operand_dims=(0, 1))
                       update_consts=() ] b c a
  in [d] }
```

**Comment by mattjj** [comment]
This one works:

```python
idx = np.arange(3)

def h(x):
  y = x * np.arange(3.)
  return np.take_along_axis(y, idx, -1)

from jax import make_jaxpr
print(make_jaxpr(h)(1.))
check_grads(h, (1.,), 1, modes=["rev"])
```

```
{ lambda b d ;  ; a.
  let c = mul a b
      e = gather[ operand_shape=(3,)
                  slice_sizes=(1,)
                  dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,)) ] c d
  in [e] }
```

So it's something about that extra dimension...

**Comment by mattjj** [comment]
This works:

```python
idx = np.arange(3).reshape(1, 3)  # ***

def h(x):
  y = x * np.arange(3.).reshape(1, 3)  # ***
  return np.take_along_axis(y, idx, -1)

from jax import make_jaxpr
print(make_jaxpr(h)(1.))
check_grads(h, (1.,), 1, modes=["rev"])
```

This doesn't:

```python
idx = np.arange(3).reshape(3, 1)  # ***

def h(x):
  y = x * np.arange(3.).reshape(3, 1)  # ***
  return np.take_along_axis(y, idx, -1)

from jax import make_jaxpr
print(make_jaxpr(h)(1.))
check_grads(h, (1.,), 1, modes=["rev"])
```

Notice the shapes `(1, 3)` vs `(3, 1)` on the marked lines.

**Comment by mattjj** [comment]
The above examples might not be good repros because they fail to evaluate when using `onp`, while the code in the OP doesn't.

**Comment by mattjj** [comment]
Hah, okay, I read @fehiepsi's OP again and I realized he basically figured out the issue for us. I was just too sloppy to read it!

I think we just need to broadcast `arr` against `indices` in our implementation of `take_along_axis`. Will attempt that now...

**Comment by mattjj** [comment]
My excuse is that I'm rusty because I've been on paternity leave for almost 3 weeks :)
